ta/pkcs11/src/processing_symm.c

```diff
@@ -751,6 +824,70 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
 
 		break;
 
+	case PKCS11_CKM_MD5_HMAC_GENERAL:
+	case PKCS11_CKM_SHA_1_HMAC_GENERAL:
+	case PKCS11_CKM_SHA224_HMAC_GENERAL:
+	case PKCS11_CKM_SHA256_HMAC_GENERAL:
+	case PKCS11_CKM_SHA384_HMAC_GENERAL:
+	case PKCS11_CKM_SHA512_HMAC_GENERAL:
+		if (session->processing->extra_ctx) {
+			hmac_len = *(uint32_t *)session->processing->extra_ctx;
+			/*
+			 * remove ptr to NW addr so that it doesn't get
+			 * freed in release_active_processing()
+			 */
+			session->processing->extra_ctx = NULL;
+
+			DMSG("hmac_len = %u", hmac_len);
+		}
+
+		switch (function) {
+		case PKCS11_FUNCTION_SIGN:
+			res = TEE_MACComputeFinal(proc->tee_op_handle,
+						  in_buf, in_size, out_buf,
+						  &out_size);
+			if (res == TEE_SUCCESS) {
+				/* truncate to hmac_len */
```

Comment:
Probably some error check if hmac_len > out_sz

Reply:
Another case could be :

User passes output buffer of size -> hmac_len (which is of size less than the expected out_sz) . However, TEE_MACComputeFinal will return short buffer error in this case in current flow. We will need to take care of this.

eg case:
For hmac_sha256_general case, caller passes hmac_len as 10 bytes and output_buffer also of out_sz 10 bytes. In the current flow, this will return error to the caller  which is not right. You might want to use your own temporary buffer here.

Reply:
For PKCS11_CKM_SHA512_HMAC_GENERAL usecases, caller can request an HMAC length of the output  in the range 0-28. In this case , if the length expected by caller is say 14 bytes, I think it should not be a short buffer error for the caller to pass an output buffer of size 14 bytes to the implementation.

Reply:
> For PKCS11_CKM_SHA512_HMAC_GENERAL usecases, caller can request an HMAC length of the output in the range 0-28. In this case , if the length expected by caller is say 14 bytes, I think it should not be a short buffer error for the caller to pass an output buffer of size 14 bytes to the implementation.

Ok, I think I get what you mean. Let me try to do some refactoring.

Reply:
@ruchi393 The refactoring turned out to be less work than previously imagined. Please take a look and comment. Thanks!

Reply:
LGTM